rfcs/0006-transaction.md

```diff
@@ -0,0 +1,299 @@
+
+```
+use rocksdb::{DB, Options, WriteBatch};
+
+let path = "_path_for_rocksdb_storage1";
+{
+    let db = DB::open_default(path).unwrap();
+    let mut batch = WriteBatch::default();
+    batch.put(b"my key", b"my value");
+    batch.put(b"key2", b"value2");
+    batch.put(b"key3", b"value3");
+    db.write(batch); // Atomically commits the batch
+}
+```
+
+You may notice that the `WriteBatch` API is very similar to the `Transaction` API we proposed below, the write operation from `WriteBatch` is also considered as atomic. The only differences are:
+
+1. the `WriteBatch` is committed by `db.write(batch)`, while the `Transaction` is committed by `txn.commit()`
+2. beside the write operations, the `Transaction` should also support the read operations, and if one key is updated in the `Transaction`, the read operation should return the updated value during the transaction.
+
+The `WriteBatch` is also considered as a prerequisite for the transaction feature, as it plays as a place to buffer the writes, and it makes MVCC possible with an iterator over it in front of the iterator over Snapshot.
+
+Now WriteBatch has been implemented in SlateDB in [this PR](https://github.com/slatedb/slatedb/pull/264). We can further add the iterator over WriteBatch in front of the iterator over Snapshot while implementing the transaction feature.
+
+### Isolation Levels
+
+When implementing the transaction feature, we should also consider the isolation level semantics. The Snapshot Isolation (SI) and Serializable Snapshot Isolation (SSI) are the two most widely used isolation levels in the industry.
+
+The Snapshot Isolation is widely used in many databases. It's gives a lightweight Snapshot on each transactions started, and checks whether the writes against the snapshot has modified by others or not during the commit. It is very efficient and is considered as a good default isolation level for most of the use cases.
+
+However, Snapshot Isolation still suffers Write Skew anomaly. For example, let's assume we have two accounts, where account A has $600, and account B has $500. Also, the business rule is that Account A's balance + Account B's balance should always be >= $200.
+
+Two transactions occur concurrently:
+
+1. Transaction 1 first checks the total amount of balance in A and B, and then transfers $550 from account A to account C.
+2. Transaction 2 first checks the total amount of balance in A and B, and then transfers $450 from account B to account D.
+
+In this case, if the two transactions are executed concurrently, the total amount of balance in A and B might be less than $200, which violates the business rule. There's no Write-Write conflict in the two transactions, but the Read-Write conflict is not detected by Snapshot Isolation.
+
+To mitigate this anomaly, we could choose:
+
+1. use Serializable Snapshot Isolation (SSI) level: increase the isolation level, and no longer allow the write skew anomaly to occur.
```

Comment:
I searched for a quick to grasp definition of SSI, and surprisingly didn't find something that is as simple as the example shared here. What do you think of expanding a bit on how SSI mitigates the read-write conflict here? *Other systems / Badger* does cover the logic. Expanding it here would be more readable.
With the sequence of RFCs added in slateDB, it might end up as a good reference material for future DB courses. I think there is benefit in making these docs easier (and quicker) to grasp.

Reply:
tuned the description here to emphasis the relationship between SSI and conflict checking, ptal 2ca5787